ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -459,7 +461,12 @@ private void readImpl(DataStreamRequestByteBuf request, ChannelHandlerContext ct
       localWrite = CompletableFuture.completedFuture(0L);
       remoteWrites = Collections.emptyList();
     } else if (request.getType() == Type.STREAM_DATA) {
-      localWrite = info.getLocal().write(request.slice(), request.getWriteOptionList(), writeExecutor);
+      if (close && request.getDataLength() == 0) {
```

Comment:
```
+    if (request.getType() == Type.STREAM_HEADER
+        || (close && request.getDataLength() == 0)) {
       localWrite = CompletableFuture.completedFuture(0L);
       remoteWrites = Collections.emptyList();
```

If it's changed here, the current server won't do the clean up, and the successor will not get clean up too.

I think the overall view looks like:
1. Server1 decides to removeStream, it will call local cleanUp, and it will send an empty close request to succssor.
2. Server2 receives the empty request, do the local cleanUp, and send the empty close request to successor.
3. Server3 receives the empty request, do the local cleanUp, since no successor here, clean up end here.

Reply:
@symious , you are right that we still need to forward the close request.  You original code is better.  How about below?

```diff
@@ -459,7 +460,11 @@ public class DataStreamManagement {
       localWrite = CompletableFuture.completedFuture(0L);
       remoteWrites = Collections.emptyList();
     } else if (request.getType() == Type.STREAM_DATA) {
-      localWrite = info.getLocal().write(request.slice(), request.getWriteOptionList(), writeExecutor);
+      if (close && request.getDataLength() == 0) {
+        localWrite = CompletableFuture.completedFuture(0L);
+      } else {
+        localWrite = info.getLocal().write(request.slice(), request.getWriteOptionList(), writeExecutor);
+      }
       remoteWrites = info.applyToRemotes(out -> out.write(request, requestExecutor));
     } else {
       throw new IllegalStateException(this + ": Unexpected type " + request.getType() + ", request=" + request);
@@ -479,7 +484,12 @@ public class DataStreamManagement {
       try {
         if (exception != null) {
           replyDataStreamException(server, exception, info.getRequest(), request, ctx);
-          removeDataStream(key, info);
+        }
+        if (close || exception != null) {
+          final StreamInfo removed = removeDataStream(key);
+          if (removed != null) {
+            Preconditions.assertSame(info, removed, "removed");
+          }
         }
       } finally {
         request.release();
```

Reply:
Updated, PTAL.